### ISSUE TITLE
[BUG] salt-pip  install M2Crypto or libvirt-python failed.  Minion-3006.7

### ISSUE THREAD

**Issue by jurajbujdoso** [saltstack/salt#66311]
**Description**
Using  salt-pip to  install M2Crypto or libvirt-python failed.

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [X] on-prem machine
- [X] VM (Virtualbox, KVM, etc. please specify)

**Steps to Reproduce the behavior**

```
salt-pip  install M2Crypto
Collecting M2Crypto
  Using cached M2Crypto-0.41.0.tar.gz (1.2 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      Error in sitecustomize; set PYTHONVERBOSE for traceback:
      AssertionError:
      Fatal Python error: init_import_site: Failed to import the site module
      Python runtime state: initialized
      Traceback (most recent call last):
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 617, in <module>
          main()
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 610, in main
          execsitecustomize()
        File "/opt/saltstack/salt/lib/python3.10/site-packages/relenv/runtime.py", line 959, in wrapper
          import sitecustomize
        File "/tmp/pip-build-env-_5mtlr_1/site/sitecustomize.py", line 22, in <module>
          assert not path in sys.path
      AssertionError
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 23.3.2 -> 24.0
```

**Expected behavior**

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

```yaml
salt-minion -V
Salt Version:
          Salt: 3006.7
 
Python Version:
        Python: 3.10.13 (main, Feb 19 2024, 03:31:20) [GCC 11.2.0]
 
Dependency Versions:
          cffi: 1.14.6
      cherrypy: 18.6.1
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.3
       libgit2: Not Installed
  looseversion: 1.0.2
      M2Crypto: 0.40.1                <--Do not be confused, this is after install of M2crypt
          Mako: Not Installed
       msgpack: 1.0.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 22.0
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.19.1
        pygit2: Not Installed
  python-gnupg: 0.4.8
        PyYAML: 6.0.1
         PyZMQ: 23.2.0
        relenv: 0.15.1
         smmap: Not Installed
       timelib: 0.2.4
       Tornado: 4.5.3
           ZMQ: 4.3.4
 
System Versions:
          dist: ubuntu 22.04.2 jammy
        locale: utf-8
       machine: x86_64
       release: 5.15.0-60-generic
        system: Linux
       version: EdgAIR OS 22.04.2 jammy

```
</details>

**Additional context**
The issue has been resolved, when I create new python virtual env independent from salt and install libvirt-python there.
After that the installation was OK.
But I was not able to install M2crypt till I install it in independent environment and continue with salt-pip install.
IMHO it could have some relation to wheel usage or something.

When running more verbose salt-pip install it provides more information:
```
/opt/saltstack/salt/salt-pip install libvirt-python -v

#This will provide more inside about the problem

note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /opt/saltstack/salt/bin/python3.10 /opt/saltstack/salt/lib/python3.10/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpkuxa3lja
  cwd: /tmp/pip-install-4ur6lpg6/libvirt-python_e5271fd61c4b4477af1b225c9ad0b5b9
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

```

```
salt-pip list
Package            Version
------------------ ---------
certifi            2023.7.22
cffi               1.14.6
charset-normalizer 3.2.0
cheroot            8.5.2
CherryPy           18.6.1
contextvars        2.4
cryptography       42.0.3
distro             1.5.0
filelock           3.13.3
idna               3.2
immutables         0.15
importlib-metadata 6.0.0
jaraco.classes     3.2.1
jaraco.collections 3.4.0
jaraco.functools   2.0
jaraco.text        3.5.1
Jinja2             3.1.3
jmespath           1.0.1
libvirt-python     10.2.0
looseversion       1.0.2
M2Crypto           0.41.0
MarkupSafe         2.1.2
more-itertools     5.0.0
msgpack            1.0.2
packaging          22.0
pip                23.3.2
portend            2.4
psutil             5.8.0
pycparser          2.21
pycryptodomex      3.19.1
pyOpenSSL          24.0.0
python-dateutil    2.8.1
python-gnupg       0.4.8
pytz               2022.1
PyYAML             6.0.1
pyzmq              23.2.0
relenv             0.15.1
requests           2.31.0
rpm-vercmp         0.1.2
salt               3006.7
setproctitle       1.3.2
setuptools         68.2.2
six                1.16.0
tempora            4.1.1
timelib            0.2.5
urllib3            1.26.18
wheel              0.42.0
xmltodict          0.13.0
zc.lockfile        1.4
zipp               3.6.0
```

**Comment by timwsuqld** [comment]
I am also having this issue on salt 3006.7 with `salt-pip install mysql`. And upgrade to 3007.0 doesn't fix anything.
```
$ salt-minion -V
Salt Version:
          Salt: 3007.0
 
Python Version:
        Python: 3.10.13 (main, Feb 19 2024, 03:31:20) [GCC 11.2.0]
 
Dependency Versions:
          cffi: 1.16.0
      cherrypy: 18.8.0
      dateutil: 2.8.2
     docker-py: 5.0.3
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.1.3
       libgit2: Not Installed
  looseversion: 1.3.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.7
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 23.1
     pycparser: 2.21
      pycrypto: Not Installed
  pycryptodome: 3.19.1
        pygit2: Not Installed
  python-gnupg: 0.5.2
        PyYAML: 6.0.1
         PyZMQ: 25.1.2
        relenv: 0.15.1
         smmap: Not Installed
       timelib: 0.3.0
       Tornado: 6.3.3
           ZMQ: 4.3.4
 
Salt Package Information:
  Package Type: onedir
 
System Versions:
          dist: ubuntu 22.04.3 jammy
        locale: utf-8
       machine: x86_64
       release: 6.5.0-1016-gcp
        system: Linux
       version: Ubuntu 22.04.3 jammy

```

**Comment by timwsuqld** [comment]
Just adding. Installing `mysql` in another python virtualenv didn't have any effect on installing it in the salt python env. I have not been able to move past this issue.

**Comment by marnovdm** [comment]
I have the same issue installing mysqlclient in 3007.0 on Debian 11:

```
root@host:~ # salt-pip install mysqlclient
Collecting mysqlclient
  Downloading mysqlclient-2.2.4.tar.gz (90 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.4/90.4 kB 4.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      Error in sitecustomize; set PYTHONVERBOSE for traceback:
      AssertionError:
      Fatal Python error: init_import_site: Failed to import the site module
      Python runtime state: initialized
      Traceback (most recent call last):
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 617, in <module>
          main()
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 610, in main
          execsitecustomize()
        File "/opt/saltstack/salt/lib/python3.10/site-packages/relenv/runtime.py", line 959, in wrapper
          import sitecustomize
        File "/tmp/pip-build-env-mcidy7d6/site/sitecustomize.py", line 22, in <module>
          assert not path in sys.path
      AssertionError
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 23.3.2 -> 24.0
[notice] To update, run: /opt/saltstack/salt/bin/python3.10 -m pip install --upgrade pip
```

**Comment by momolemo** [comment]
seems similar to https://github.com/saltstack/salt/issues/65980 and https://github.com/saltstack/salt/issues/65972
TLDR => workaround : 
```
salt-pip install pip==22.3.1
salt-pip install mysqlclient
```
source :
[salt-pip install ](https://github.com/saltstack/salt/issues/65980#issuecomment-1991357318)

**Comment by luckylinux** [comment]
My "solution" mentioned in the other post is to apply a patch `salt-pip` itself:
https://github.com/saltstack/salt/issues/65980#issuecomment-2084731676

Patch and Instructions (as well as a LONG Explanation / Analysis / Test) is provided in my comment ABOVE the mentioned link.

**Comment by smarsching** [comment]
For those who are interested, I was able to track this down to an issue in relenv: saltstack/relenv#182

**Comment by joachimtingvold** [comment]
I stumbled upon a flavor of this issue, combined with #35013 and #57339. I had upgraded salt-minions from 3000.x to 3007.x, and anywhere the x509 module was used (v1, not the new v2), I got the `'x509' __virtual__ returned False: Could not load x509 state: m2crypto unavailable` error.

This was on Debian 11/12, already having `python3-m2crypto` installed. Trying to install it via `salt-pip install m2crypto` gave me the same error as experienced in this issue. There was also a few other dependencies that needed to be installed.

Final solution for me was;

```
salt-pip install pip==22.3.1
apt install libssl-dev swig python3-dev gcc
salt-pip install m2crypto
```

... or if you want to fix it via salt-master;


```
x509_fix_pip_module:
  cmd.run:
    - name: |
        salt-pip install pip==22.3.1
        apt install -y libssl-dev swig python3-dev gcc
        salt-pip install m2crypto
    - shell: /bin/bash
```

**Comment by dmurphy18** [comment]
@dwoz Confirmed this is real with Ubuntu 22.04 on an AWS VM with Salt 3006.8, but checking libvirt on PyPI, 
```
Requirements
To build the libvirt python binary requires the GCC and pkg-config tools, and development headers for the libvirt and python libraries.
```

So after installing the following:
```
apt install build-essential
apt install pkg-config
apt install libvirt-dev
```

The command ```salt-pip install libvirt``` and the result is successful
```
root@ip-172-31-2-109:/home/ubuntu# salt-pip list  | grep libvirt
libvirt-python     10.5.0

[notice] A new release of pip is available: 23.3.2 -> 24.2
[notice] To update, run: /opt/saltstack/salt/bin/python3.10 -m pip install --upgrade pip
root@ip-172-31-2-109:/home/ubuntu# 
```

Seems it is working similar to regular pip, which salt-pip is working similarly,salt-pip and pip3 both need the requirements to install libvirt satisfied before the install can be successful.

Closing this since functioning as expected. Feel free to reopen if there is other information pertinent to the issue.
Note: salt-pip duplicates pip3 functionality, pip installing the into the correct place for Salt onedir architecture, and has the same limitations as pip3, that is, if pip3 has requirements, then salt-pip has similar requirements to function correctly.

**Comment by jamest-pin** [comment]
I can't see how this is working as expected.

Still getting this issue on salt 3006 and 3007, latest minor versions, fresh install. To reproduce, 
Build the container e.g. `docker build --build-arg SALT_VERSION=3006 --build-arg UBUNTU_VERSION=22.04`

(note, `apt install -y libssl-dev swig python3-dev gcc software-properties-common build-essential pkg-config libvirt-dev` is required. I don't know which particular packages are required, this is just the list I've compiled over time which fixes most issues).

1. Try to run a command like `salt-pip install m2crypto`.
2. The command fails.
4. Use `/opt/saltstack/salt/bin/python3.10 -m pip install m2crypto`
5. The command now works. 

here's my Dockerfile
```Dockerfile
# Dockerfile for a salt master
# The ubuntu version to use. Specify this at cli with --build-arg UBUNTU_VERSION=xx.xx
ARG UBUNTU_VERSION
FROM ubuntu:${UBUNTU_VERSION}

# Whether to install salt-master or salt-minion
ARG SALT_STATUS
# Version of salt to install
ARG SALT_MAJOR_VERSION
RUN apt update
# install curl which is needed below
# -y means don't prompt for confirmation
RUN apt install -y curl

##
# Pin salt version
##
# Avoid interactive prompt when tzdata gets installed by salt-minion
ENV TZ=Australia/Perth

# OS dependenceis required for things like salt-pip install m2crypto to work
RUN apt install -y libssl-dev swig python3-dev gcc software-properties-common build-essential pkg-config libvirt-dev

# Add salt apt repo
# using . instead of source because source is bash specific and results in an error in the build
# mkdir /etc/apt/keyrings is part of salt instructions for OS prior to 22.04 (https://docs.saltproject.io/salt/install-guide/en/latest/topics/install-by-operating-system/ubuntu.html)
# If mkdir /etc/apt/keyrings is not run, in 20.04 (and presumably below) you get curl: (23) Failed writing body (0 != 1769)
# the -p (aka create intermediate directories) is used because it doesn't error if folder already exists
RUN mkdir -p /etc/apt/keyrings
 # Download public key
RUN curl -fsSL https://packages.broadcom.com/artifactory/api/security/keypair/SaltProjectKey/public | tee /etc/apt/keyrings/salt-archive-keyring.pgp \
# Create apt repo target configuration
  && curl -fsSL https://github.com/saltstack/salt-install-guide/releases/latest/download/salt.sources | tee /etc/apt/sources.list.d/salt.sources
RUN apt update

# Pin salt to correct version
# COPY <<-"EOT" /etc/apt/preferences.d/salt-pin-1001
# Package: salt-*
# Pin: version ${SALT_MAJOR_VERSION}.*
# Pin-Priority: 1001
# EOT
RUN printenv > /printenv
RUN echo Package: salt-* | tee /etc/apt/preferences.d/salt-pin-1001
RUN echo Pin: version "${SALT_MAJOR_VERSION}"'.*' | tee -a /etc/apt/preferences.d/salt-pin-1001
RUN echo Pin-Priority: 1001 | tee -a /etc/apt/preferences.d/salt-pin-1001

# Install salt master
# https://docs.saltproject.io/salt/install-guide/en/latest/topics/bootstrap.html
RUN apt install -y salt-master salt-minion salt-common

# Set default log level. Gets overridden by docker compose usually.
# log level https://docs.saltproject.io/en/latest/ref/configuration/logging/index.html
ENV SALT_LOG_LEVEL=info

# Make salt-master run on startup and not exit by calling it as an executable instead of
# starting the service. Note: this (intentionally) makes the containers stop if the salt
# minion process exits for any reason including errors or timeout.
# To get into the container in a shell, simple run docker run -it <IMAGE_ID> sh
RUN echo "salt-master --log-level=\${SALT_LOG_LEVEL:-warning}" > /entrypoint.sh \
  && chmod +x entrypoint.sh
ENTRYPOINT [ "sh", "-c", "/entrypoint.sh" ]
```

Here are the logs from the pip commands
```shell
root@fb9e6f8ef81c:/# salt-pip install m2crypto
Collecting m2crypto
  Downloading m2crypto-0.44.0.tar.gz (363 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.8/363.8 kB 39.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      Error in sitecustomize; set PYTHONVERBOSE for traceback:
      AssertionError:
      Fatal Python error: init_import_site: Failed to import the site module
      Python runtime state: initialized
      Traceback (most recent call last):
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 627, in <module>
          main()
        File "/opt/saltstack/salt/lib/python3.10/site.py", line 620, in main
          execsitecustomize()
        File "/opt/saltstack/salt/lib/python3.10/site-packages/relenv/runtime.py", line 969, in wrapper
          import sitecustomize
        File "/tmp/pip-build-env-957pgpaf/site/sitecustomize.py", line 22, in <module>
          assert not path in sys.path
      AssertionError
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

[notice] A new release of pip is available: 23.3.2 -> 25.0.1
[notice] To update, run: /opt/saltstack/salt/bin/python3.10 -m pip install --upgrade pip
root@fb9e6f8ef81c:/# /opt/saltstack/salt/bin/python3.10 -m pip install m2crypto
Collecting m2crypto
  Using cached m2crypto-0.44.0.tar.gz (363 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: m2crypto
  Building wheel for m2crypto (pyproject.toml) ... done
  Created wheel for m2crypto: filename=M2Crypto-0.44.0-cp310-cp310-linux_aarch64.whl size=747712 sha256=85fa4c73a473cde0ff189d099b62dcdadc6a760ea1591a508c4f93460e8cac99
  Stored in directory: /root/.cache/pip/wheels/90/02/b6/975e686438122bd0d20a7892f01a9a1bc7d76f745c9e172e06
Successfully built m2crypto
Installing collected packages: m2crypto
  Do not adjust rpath of /opt/saltstack/salt/lib/python3.10/site-packages/M2Crypto/_m2crypto.cpython-310-aarch64-linux-gnu.so
Successfully installed m2crypto-0.44.0
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv

[notice] A new release of pip is available: 23.3.2 -> 25.0.1
[notice] To update, run: /opt/saltstack/salt/bin/python3.10 -m pip install --upgrade pip
root@fb9e6f8ef81c:/#
```
